lib/libzpool/zrepl_mgmt.c

```diff
@@ -202,3 +247,46 @@ uzfs_zinfo_free(zvol_info_t *zinfo)
 	free(zinfo);
 	return (0);
 }
+
+void
+uzfs_zvol_get_last_committed_io_no(zvol_info_t *zinfo, uint64_t *io_seq)
+{
+	uzfs_zap_kv_t zap;
+	zap.key = "io_seq";
+	zap.value = 0;
+	zap.size = sizeof (*io_seq);
+
+	uzfs_read_zap_entry(zinfo->zv, &zap);
+	*io_seq = zap.value;
+}
+
+void
+uzfs_zvol_store_last_committed_io_no(zvol_state_t *zv, uint64_t io_seq)
+{
+	uzfs_zap_kv_t *kv_array[0];
+	uzfs_zap_kv_t zap;
+	zap.key = "io_seq";
+	zap.value = io_seq;
+	zap.size = sizeof (io_seq);
+
+	kv_array[0] = &zap;
+	VERIFY0(uzfs_update_zap_entries(zv,
+	    (const uzfs_zap_kv_t **) kv_array, 1));
+}
+
+void
+uzfs_zinfo_update_io_seq_for_all_volumes(void)
+{
+	zvol_info_t *zinfo;
+	(void) pthread_mutex_lock(&zvol_list_mutex);
+	SLIST_FOREACH(zinfo, &zvol_list, zinfo_next) {
+		if (zinfo != NULL) {
+			/* Take refcount */
+			uzfs_zinfo_take_refcnt(zinfo, B_TRUE);
```

Comment:
we don't need to take ref here as we hold zvol_list_mutex thus nobody should be able to remove a zinfo from list (or destroy it).

Reply:
True. But to be candid I did not like this idea of helding global lock for a longer time. As the number of volumes will increased on storage box, it will be a bottleneck to system in every 10mint(or for an interval). Refcount need a change, I want to change this zinfo refcount or want to make it little lighter (by avoidning such a coarse lock.